### PR TITLE
Add Uri subclass to avoid dynamic properties

### PR DESCRIPTION
**Pull request by markstory**
Refs https://github.com/cakephp/cakephp/issues/16158

Remove the usage of dynamic properties on Uri by subclassing and making the properties concrete. I'm not sure I like using the builder methods to reconstruct the URI. Another solution I might explore is to create proxy methods, but first I want to profile the runtime impacts of this change.

### CODE REVIEW

File: src/Http/Uri.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    /**
+     * Backwards compatible property read
+     *
+     * @param string $prop The property to read.
+     * @return string|null
+     */
+    public function __get($prop): ?string
+    {
+        if ($prop !== 'base' && $prop !== 'webroot') {
+            return null;
```
Comment (othercorey) [inline src/Http/Uri.php]: Is it convention to call trigger_error() here?

Comment (ADmad) [reply src/Http/Uri.php]: trigger_error() would allow us to show a better error message.

Comment (othercorey) [reply src/Http/Uri.php]: This is used the php docs example:
```php
        trigger_error(
            'Undefined property via __get(): ' . $name .
            ' in ' . $trace[0]['file'] .
            ' on line ' . $trace[0]['line'],
            E_USER_NOTICE);
```

**Comment by othercorey** [comment]
The memcached error is due to invalid parameter parsing in the extension. It was patched, but they haven't released for 2 years.  

The error went away so I assumed setup-php was building from git, but guess not.

File: src/Http/Uri.php (hunk 1 above)
Comment (markstory) [reply src/Http/Uri.php]: Yeah I'm not digging how this interface is working. I'm tempted to introduce a container for the Uri & our custom attributes and extract the Uri in the request. My hunch is that very few or no userland code is mucking with the guts of building a request.

**Comment by markstory** [comment]
I rejigged this to use a decorator instead of a subclass. I think it will add less overhead than copying all the data into the subclass.

**Comment by ADmad** [comment]
LGTM, just the static analyzer errors need to be sorted out.

File: src/Http/Uri.php [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.3.0
```
Comment (saeideng) [inline src/Http/Uri.php]:
```suggestion
 * @since         4.4.0
```

also in `UriTest.php ` file

File: src/Http/Uri.php [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+    /**
+     * @var \Psr\Http\Message\UriInterface
+     */
+    private $uri;
+
+    /**
+     * Constructor
+     *
+     * @param \Psr\Http\Message\Uri $uri Uri instance to decorate
```
Comment (ADmad) [inline src/Http/Uri.php]:
```suggestion
     * @param \Psr\Http\Message\UriInterface $uri Uri instance to decorate
```

File: src/Http/Uri.php [hunk 4]
```diff
@@ -158,55 +158,55 @@ public function getFragment()
      */
     public function withScheme($scheme)
     {
-        return $this->uri->withScheme($scheme);
+        return new static($this->uri->withScheme($scheme), $this->base, $this->webroot);
```
Comment (ADmad) [inline src/Http/Uri.php]: Swapped one error for another 😆. Static analysers are not easy to please.

**Pull request merged**